src/main/kotlin/org/rust/ide/intentions/InvertIfIntention.kt

```diff
@@ -0,0 +1,53 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.intentions
+
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import org.rust.ide.utils.simplifyNegatedExpression
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.ancestorStrict
+import org.rust.lang.core.psi.ext.operatorType
+import org.rust.lang.utils.negate
+
+class InvertIfIntention : RsElementBaseIntentionAction<RsIfExpr>() {
+    override fun findApplicableContext(project: Project, editor: Editor, element: PsiElement): RsIfExpr? {
+        val ifStatement = element.ancestorStrict<RsIfExpr>() ?: return null
+
+        return if (ifStatement.elseBranch?.block != null &&
+            ifStatement.block != null &&
+            checkIsSuitableCondition(ifStatement)) ifStatement else null
+    }
+
+    override fun invoke(project: Project, editor: Editor, ctx: RsIfExpr) {
```

Comment:
IMO now it's better to call `DemorgansLawIntention` directly like this (note that `ifExpr` can be replaced before `DemorgansLawIntention` call):
```Kotlin
    override fun invoke(project: Project, editor: Editor, ctx: Context) {
        val negatedCondition = ctx.condition.expr.negate() as? RsExpr ?: return

        val newIf = RsPsiFactory(project).createIfElseExpression(negatedCondition, ctx.elseBlock, ctx.thenBlock)
        ctx.ifExpr.replace(newIf)

        val demorgansLawIntention = DemorgansLawIntention()

        if (negatedCondition is RsUnaryExpr && negatedCondition.excl != null) {
            val binaryExpr = (negatedCondition.expr as? RsParenExpr)?.expr as? RsBinaryExpr ?: return
            val binaryOp = binaryExpr.binaryOp

            if (demorgansLawIntention.isAvailable(project, editor, binaryOp)) {
                demorgansLawIntention.invoke(
                    project,
                    editor,
                    DemorgansLawIntention.Context(binaryExpr, binaryOp.operatorType)
                )
            }
        }
    }
```
Currently `DemorgansLawIntention` is too complicated to be extracted: just look at `simplifyNegatedExpression`, `convertLeafExpression`, `convertConjunctionExpression` inside `ExprUtils.kt`. So it should be refactored and extracted in the future.

Reply:
Ooh. OK. But can I copypaste or something like this ```convertLeafExpression```? Or is it fine to leave ```!(2 == 2)``` not simplified?

Reply:
Ah, you're right, it doesn't work in this case. The problem relates to `DemorgansLawIntention` itself, espessially with this line inside `findApplicableContext`:
`val binExpr = element.ancestorStrict<RsBinaryExpr>() ?: return null`

It searches strict ancestor with type `RsBinaryExpr`, so you have to pass something inside binary expr to the intention).

You can use code like this here (`isAvailable` call is redundant is you call `invoke(project: Project, editor: Editor, element: PsiElement)`):

```Kotlin
override fun invoke(project: Project, editor: Editor, ctx: Context) {
    val negatedCondition = ctx.condition.expr.negate() as? RsExpr ?: return

    val ifExpr = RsPsiFactory(project).createIfElseExpression(negatedCondition, ctx.elseBlock, ctx.thenBlock)
    val replacedIfExpr = ctx.ifExpr.replace(ifExpr) as RsIfExpr
    val newCondition = replacedIfExpr.condition!!.expr

    if (newCondition is RsUnaryExpr && newCondition.excl != null) {
        val binaryExpr = (newCondition.expr as? RsParenExpr)?.expr as? RsBinaryExpr ?: return
        DemorgansLawIntention().invoke(project, editor, binaryExpr.binaryOp)
    }
}
```

Reply:
Will it work with 2 == 2? As I understand, DemorganLaw looks for && or ||.

Reply:
You're right, it works only for binary expressions with logical operators at top-level (like `2 == 2 && 3 == 3`, but not for `2 == 2`). And it checks inside `DemorgansLawIntention.findApplicableContext` when you call `invoke`.

Reply:
And do I leave !(2 == 2) without changes now?

Reply:
Seems like it can be fixed via improving `PsiElement.negate()` (and the problem is not about your intention only, but affects other code insight related to boolean expressions). I think you can leave it as is for now and focus on the intention. I'll fix it later.

Reply:
Yes!. I found it in Kotlin plugin and want to write about it too.

Reply:
There are no ```createIfElseExpression```, only IfExpression with one branch. Can I add it?

Reply:
Yes, something like
```Kotlin
    fun createIfElseExpression(condition: RsExpr, thenBlock: RsBlock, elseBlock: RsBlock): RsIfExpr {
        val result = createExpressionOfType<RsIfExpr>("if ${condition.text} { () } else { () }")
        result.block!!.replace(thenBlock)
        result.elseBranch!!.block!!.replace(elseBlock)
        return result
    }
```

Reply:
Is there ```!!``` operator instead of ```?.``` for throwing exception if something is broken in ```createExpressionOfType``` function?

Reply:
We have valid `condition`, so `createExpressionOfType<RsIfExpr>("if ${condition.text} { () } else { () }")` must be OK.

Reply:
So, it doesn't matter ```?.``` or ```!!``` in this case?

Reply:
I'm just learning Kotlin, so I think about difference of this operators in this case. In my opinion, it is a idealogic problem, based on what we want. If we want this function is working in all cases, also in case, when someone broke ```createExpressionOfType```. Or we want to throw exception, when someone broke this function.

Reply:
OK! I got it.